### ISSUE TITLE
Fitting validation

### ISSUE THREAD

**Issue by cklyne** [applied-ml-bde/bde#33]
## Idea
We should implement validation steps during training.
We've already implemented corresponding attributes in the APIs of the estimator classes and the training algorithms, but they are currently unused.
I propose implementing the following:
 - In the estimator classes when the training data is prepared it should be split into 2 training sets based on the validation attribute (currently all of the data is used for a training set, and an empty validation set is created).
 - In the jitted training methods, after evaluating the loss and metrics over the training data, the same evaluation should be done in exactly the same order on the validation data, without optimizing the parameters. The results should be appended to the history attribute.
